### DISCUSSION TITLE
DuckDB DBGEN proposed extensions

### DISCUSSION THREAD

**Discussion by pedroerp** [duckdb/duckdb#3657] [Ideas]
Greetings! We recently started using DuckDB's dbgen in Velox to generate TPC-H data on-the-fly, but we bumped into a few API limitations. Before going ahead and forking the code within our project, I'd like to check with the DuckDB community if these limitations are worth adding to DuckDB directly, and whether the APIs we are currently using are supposed to be used in that way. 

The main goal of the project is to use dbgen (for now DuckDB's dbgen) to generate data using Velox's data structures, in such a way that data can be dynamically generated at query time. The main API being used are the mk_order, mk_part, etc, functions provided by dbgen.

1. The first issue is related the global variables used by dbgen C codebase, what makes the code not thread-safe. To go around this problem, we currently lock concurrent readers (similar to the approach in DuckDB's dbgen.cpp), but the approach is obviously not scalable. To remove the limitation, the shared global structures could be moved to a context object passed to these calls, or turning these global variables thread_local. Is it something DuckDB would be interested in extending? I understand that the fact that dbgen is supposed to be hermetically-sealed environment is a conflicting goal in this case. 

2. In order to support parallel generation of different splits, Velox needs to support deterministic generation of TPC-H datasets given a row offset. Even though this mostly works today, in order to make the random number generation deterministic given an offset, it requires the speed_seed.cpp functions to be exposed as part of the API (in that case the sd_*() functions signatures would need to also be part of dsstypes.h). Additionally, some sd_* functions are missing and would have to be added (sd_region()) is one example. 

Given the context, would these changes be of interest of the DuckDB community?

**Comment by Mytherin** [comment]
Thanks for opening up the discussion!

Yes, all of those changes are for sure interesting to us. We run dbgen relatively frequently, and adding support for parallel generation would be fantastic. More than happy to review a PR on this!

As for generating Velox's data structures natively - there are only a few DuckDB specific methods used  (`append_value`, `append_string`, etc). It should be possible to make e.g. a virtual class that both DuckDB and Velox can override, which would allow the bulk of the code-base to be shared between the two projects.

**Comment by pedroerp** [comment]
Thanks, I'll be looking into these moving forward (I was out for a few weeks). 

The first PR that exposes functions to allow generation given a certain offset is out: #3775
